### ISSUE TITLE
okta_profile_mapping fails to set userName override in oauth app

### ISSUE THREAD

**Issue by dkulchinsky** [okta/terraform-provider-okta#949]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

0.14.7

### Affected Resource(s)

- okta_profile_mapping

### Terraform Configuration Files

```hcl
resource "okta_app_oauth" "oidc_app" {
  label                  = var.app_name
  type                   = var.app_type
  omit_secret            = true
  grant_types            = var.grant_types
  redirect_uris          = var.redirect_uris
  login_mode             = var.login_mode
  login_uri              = var.login_uri
  response_types         = var.response_types
  hide_ios               = var.hide_ios
  hide_web               = var.hide_web
  refresh_token_rotation = "STATIC"
  groups_claim {
    type        = "FILTER"
    filter_type = "REGEX"
    name        = "groups"
    value       = var.app_groups_claim_regex
  }

  // users and/or groups are assigned through okta_app_group_assignment resource
  lifecycle {
    ignore_changes = [users, groups]
  }
}

data "okta_user_profile_mapping_source" "user" {}

resource "okta_profile_mapping" "username" {
  source_id          = "data.okta_user_profile_mapping_source.user.id"
  target_id          = "okta_app_oauth.oidc_app.id"
  delete_when_absent = false
  always_apply       = true

  mappings {
    id          = "userName"
    expression  = "String.substringBefore(user.email, '@')"
    push_status = "PUSH"
  }
}
```

### Expected Behavior

I'm expecting this code to generate a mapping override like the one I can create via the console:
![image](https://user-images.githubusercontent.com/61992380/152588912-32e96e0b-1ced-49e6-ae31-341818297260.png)

### Actual Behavior

Plan seems correct, however it fails during Apply with:

```
Error: failed to get profile mapping: the API returned an error: The request was not valid: Invalid source or target id. Source or target id must be of either App Instance or UserType id, Status: 400 Bad Request

  on modules/oidc_app/main.tf line 28, in resource "okta_profile_mapping" "username":
  28: resource "okta_profile_mapping" "username" {
```

### Steps to Reproduce

1. use the TF code I provided above
2. `terrafirm plan` <-- should work fine
2. `terraform apply` <-- fails with above error

### Important Factoids

I looked at the existing examples and related issues, they seem to be mapping in the other direction, however as I understand based on the console screenshot that I pasted the `source_id` should point to `data.okta_user_profile_mapping_source.user.id` (for the User's base profile) and the `target_id` should point to the `okta_app_oauth.id` attribute.

I also queried the API to look for clues and I saw that in the App that I was setting manually through the UI, there are these attributes:
```json
  "credentials": {
    "userNameTemplate": {
      "template": "String.substringBefore(user.email, '@')",
      "type": "CUSTOM",
      "pushStatus": "PUSH"
    },
```

and I found that these are indeed available as attributes in the `okta_app_oauth` resource, so I tried that:
```hcl
  user_name_template             = "String.substringBefore(user.email, '@')"
  user_name_template_type        = "CUSTOM"
  user_name_template_push_status = "PUSH"
```

this seem to have done the right thing, however when logging in to the app it doesn't work, I get the full email in the preferred_username claim, but the same works just fine when configuring through the UI.

perhaps I'm missing some publish operation/step?

I feel like I'm missing something fundamental so any advice here would be really appreciated!

**Comment by dkulchinsky** [comment]
I'm reading this Okta KB note: https://support.okta.com/help/s/article/Application-Usernames-are-not-being-updated-automatically?language=en_US

and I believe my attempt to set the `user_name_template` meets one of these requirements:
 ```
Application Username format is set to Custom and is not using deprecated expressions
```

I'm indeed using a `CUSTOM` format, and the expression is not a deprecated one.

I think what is missing is the `apply` part?

when making this change from the UI, when saving the mapping it asks this question:
![image](https://user-images.githubusercontent.com/61992380/152611050-beb189c1-9df9-4ec1-8b49-d23df94ce52e.png)

perhaps this step is missing from the provider's code?

**Comment by monde** [comment]
hi @dkulchinsky we'll be looking in to this next week. Thanks for all the details, that is really helpful.

**Comment by dkulchinsky** [comment]
Hey @monde!

Thanks for replying 🙂 

Let me know if you need any additional information, and sorry if I'm confusing topics and resources, I assumed at first this is done with okta_profile_mapping but then saw it's actually attributes in the okta_app_oauth but doesn't seem to work like it does in the UI.

**Comment by dkulchinsky** [comment]
I've managed to pull down the mapping (after creating it via the UI) for the application from the API:

```json
{
  "id": "prmab69ciJ9tRatSD5d6",
  "source": {
    "id": "oty3uvnbk9sdp9ZYW5d6",
    "name": "user",
    "type": "user",
    "_links": {
      "self": {
        "href": "https://<domain>.okta.com/api/v1/meta/types/user/oty3uvnbk9sdp9ZYW5d6"
      },
      "schema": {
        "href": "https://<domain>.okta.com/api/v1/meta/schemas/user/osc3uvnbk9sdp9ZYW5d6"
      }
    }
  },
  "target": {
    "id": "0oaab69c5pIjg2u4d5d6",
    "name": "oidc_client",
    "type": "appuser",
    "_links": {
      "self": {
        "href": "https://<domain>.okta.com/api/v1/apps/0oaab69c5pIjg2u4d5d6"
      },
      "schema": {
        "href": "https://<domain>.okta.com/api/v1/meta/schemas/apps/0oaab69c5pIjg2u4d5d6/default"
      }
    }
  },
  "properties": {
    "userName": {
      "expression": "String.substringBefore(user.email, \"@\")",
      "pushStatus": "PUSH"
    },
    "name": {
      "expression": "user.displayName",
      "pushStatus": "PUSH"
    },
    "nickname": {
      "expression": "user.nickName",
      "pushStatus": "PUSH"
    },
    "given_name": {
      "expression": "user.firstName",
      "pushStatus": "PUSH"
    },
    "middle_name": {
      "expression": "user.middleName",
      "pushStatus": "PUSH"
    },
    "family_name": {
      "expression": "user.lastName",
      "pushStatus": "PUSH"
    },
    "email": {
      "expression": "user.email",
      "pushStatus": "PUSH"
    },
    "profile": {
      "expression": "user.profileUrl",
      "pushStatus": "PUSH"
    },
    "zoneinfo": {
      "expression": "user.timezone == null ? \"America/Los_Angeles\" : user.timezone",
      "pushStatus": "PUSH"
    },
    "locale": {
      "expression": "user.countryCode == null ? \"en-US\" : (String.stringContains(user.countryCode, \"_\") ? String.substringBefore(user.countryCode, \"_\") + \"-\" + String.substringAfter(user.countryCode, \"_\") : user.countryCode)",
      "pushStatus": "PUSH"
    },
    "phone_number": {
      "expression": "user.primaryPhone",
      "pushStatus": "PUSH"
    },
    "street_address": {
      "expression": "user.streetAddress",
      "pushStatus": "PUSH"
    },
    "locality": {
      "expression": "user.city",
      "pushStatus": "PUSH"
    },
    "region": {
      "expression": "user.state",
      "pushStatus": "PUSH"
    },
    "postal_code": {
      "expression": "user.zipCode",
      "pushStatus": "PUSH"
    },
    "country": {
      "expression": "user.countryCode",
      "pushStatus": "PUSH"
    }
  },
  "_links": {
    "self": {
      "href": "https://<domain>.okta.com/api/v1/mappings/prmab69ciJ9tRatSD5d6"
    }
  }
}
```

based on above, my attempt with the `okta_profile_mapping` resource should be correct:

```hcl
data "okta_user_profile_mapping_source" "user" {}

resource "okta_profile_mapping" "username" {
  source_id          = "data.okta_user_profile_mapping_source.user.id"
  target_id          = "okta_app_oauth.oidc_app.id"
  delete_when_absent = false
  always_apply       = true

  mappings {
    id          = "userName"
    expression  = "String.substringBefore(user.email, '@')"
    push_status = "PUSH"
  }
}
```

the `source_id` comes from `data.okta_user_profile_mapping_source.user.id`, which has the following attributes (pulled from the state) that match the above API response:
```json
      "mode": "data",
      "type": "okta_user_profile_mapping_source",
      "name": "user",
          "attributes": {
            "id": "oty3uvnbk9sdp9ZYW5d6",
            "name": "user",
            "type": "user"
```

the `target_id` is the `okta_app_oauth.oidc_app.id`, which has the following value:
```json
      "mode": "managed",
      "type": "okta_app_oauth",
      "name": "oidc_app",
      "provider": "provider[\"registry.terraform.io/okta/okta\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "0oaab69c5pIjg2u4d5d6",
```

Based on the error message, I believe the failing request is here: https://github.com/okta/terraform-provider-okta/blob/c76b5e698308507044f1dd3385435f8d1c6a2c05/sdk/mappings.go#L33

I ran this query myself using the above source/target IDs, and it works:
```
❯ curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://<domain>.okta.com/api/v1/mappings?sourceId=oty3uvnbk9sdp9ZYW5d6&targetId=0oaab69c5pIjg2u4d5d6" | jq .

[
  {
    "id": "prmab69ciJ9tRatSD5d6",
    "source": {
      "id": "oty3uvnbk9sdp9ZYW5d6",
      "name": "user",
      "type": "user",
      "_links": {
        "self": {
          "href": "https://<domain>.okta.com/api/v1/meta/types/user/oty3uvnbk9sdp9ZYW5d6"
        },
        "schema": {
          "href": "https://<domain>.okta.com/api/v1/meta/schemas/user/osc3uvnbk9sdp9ZYW5d6"
        }
      }
    },
    "target": {
      "id": "0oaab69c5pIjg2u4d5d6",
      "name": "oidc_client",
      "type": "appuser",
      "_links": {
        "self": {
          "href": "https://<domain>.okta.com/api/v1/apps/0oaab69c5pIjg2u4d5d6"
        },
        "schema": {
          "href": "https://<domain>.okta.com/api/v1/meta/schemas/apps/0oaab69c5pIjg2u4d5d6/default"
        }
      }
    },
    "_links": {
      "self": {
        "href": "https://<domain>.okta.com/api/v1/mappings/prmab69ciJ9tRatSD5d6"
      }
    }
  }
]
```

so, this is somewhat buffeling and hope this extra info helps :)

**Comment by monde** [comment]
@dkulchinsky can you try using the `always_apply` argument (this implies the Apply button seen your screen shot) and see if it works for you. See https://github.com/okta/terraform-provider-okta/blob/master/website/docs/r/profile_mapping.html.markdown#argument-reference

I will investigate further if this doesn't work for you. Let me know.

**Comment by dkulchinsky** [comment]
> @dkulchinsky can you try using the `always_apply` argument (this implies the Apply button seen your screen shot) and see if it works for you. See [`master`/website/docs/r/profile_mapping.html.markdown#argument-reference](https://github.com/okta/terraform-provider-okta/blob/master/website/docs/r/profile_mapping.html.markdown#argument-reference)
> 
> I will investigate further if this doesn't work for you. Let me know.

Hey @monde! thanks for taking a look.

I do have `always_apply = true` in my configuration, as per my original comment, this is my config:
```hcl
data "okta_user_profile_mapping_source" "user" {}

resource "okta_profile_mapping" "username" {
  source_id          = "data.okta_user_profile_mapping_source.user.id"
  target_id          = "okta_app_oauth.oidc_app.id"
  delete_when_absent = false
  always_apply       = true

  mappings {
    id          = "userName"
    expression  = "String.substringBefore(user.email, '@')"
    push_status = "PUSH"
  }
}
```

but this fails with:
```
Error: failed to get profile mapping: the API returned an error: The request was not valid: Invalid source or target id. Source or target id must be of either App Instance or UserType id, Status: 400 Bad Request

  on modules/oidc_app/main.tf line 28, in resource "okta_profile_mapping" "username":
  28: resource "okta_profile_mapping" "username" {
```

since the above is failing, I tried to use the following attributes in the `okta_app_oauth` resource instead of using `okta_profile_mapping`:
```hcl
  user_name_template             = "String.substringBefore(user.email, '@')"
  user_name_template_type        = "CUSTOM"
  user_name_template_push_status = "PUSH"
```

but this seem to have no effect and I was getting still `<name>@<domain>` in the `preferred_username` claim instead of `<name>`

**Comment by dkulchinsky** [comment]
@monde I guess my main confusion is what method should I use to set this up?

1. if I should use the `okta_profile_mapping` resource - then it fails with the error I mentioned
2. if I should set this with `user_name_template*` attributes in the `okta_app_oauth` resources - then that doesn't seem to take effect once applied

**Comment by bogdanprodanj** [comment]
@dkulchinsky I think your problem is trivial. This part of the script is treated by the terraform as plain strings:
```
  source_id = "data.okta_user_profile_mapping_source.user.id"
  target_id = "okta_app_oauth.oidc_app.id"
```
You should drop the quotes:
```
  source_id =  data.okta_user_profile_mapping_source.user.id
  target_id =  okta_app_oauth.oidc_app.id
```

Another thing you should do in the `okta_app_oauth` resource is to ignore the changes for username templates:
```
lifecycle {
    ignore_changes = [user_name_template, user_name_template_type, user_name_template_push_status]
  }
```

**Comment by dkulchinsky** [comment]
@bogdanprodanj @monde folks, I feel quite silly now 🤦🏼‍♂️ 

what a simple oversight... darn it.... I'm sorry for wasting your time on this.

I think I carried over this when using the example in the doc here: https://registry.terraform.io/providers/okta/okta/latest/docs/resources/profile_mapping, it uses the pre TF 0.12 interpolation syntax and had quotes, I used that as a source and just kept the quotes I guess.

appreciate your help on this folks! after removing the quotes everything works as expected.
